src/cargo/util/toml.rs

```diff
@@ -540,6 +540,18 @@ impl TomlManifest {
             }
         }
 
+        {
+            let mut names_sources = HashMap::new();
+            for dep in deps.iter() {
+                let name = dep.name();
+                let sources = names_sources.entry(name).or_insert(HashSet::new());
+                if sources.insert(dep.source_id()) && sources.len() > 1 {
```

Comment:
Oh, right! In that case we don't need the value here to be a `HashSet<SourceId>`, right? If it's just a `SourceId` we could just compare against what was already there.